src/datasets/arrow_dataset.py

```diff
@@ -2175,6 +2178,50 @@ def generator():
         logger.info(f"Finished writing TFRecord to {filename}")
         self = None  # delete the dataset reference used by tf_dataset
 
+    def _write_csv(self, file_obj: BinaryIO, batch_size: int, **to_csv_kwargs) -> int:
+        """
+        Writes the pyarrow table as CSV to a binary file handle.
+        Caller is responsible for opening and closing the handle.
+        """
+        written = 0
+        to_csv_kwargs.pop("header", None)
+        to_csv_kwargs.pop("path_or_buf", None)
+        to_csv_kwargs.pop("encoding", None)
+
+        for offset in range(0, self._data.num_rows, batch_size):
+            batch = self._data.slice(offset, offset + batch_size)
+            csv_str = batch.to_pandas().to_csv(
```

Comment:
Also note that we have custom features Array2D, Array3D etc. defined in features.py

They correspond to the custom pyarrow types: Array2DExtensionType, Array3DExtensionType etc.
We also have their corresponding custom pandas type PandasArrayExtensionDType

The conversion using to_pandas() for these types is supported since we have the associated pandas type, but I don't know how pandas supports custom types when converting to csv.

I just tested and this code crashes unfortutanely:
```python
from datasets import Dataset, Features, Array2D

features = Features({"a": Array2D(dtype="int32", shape=(2, 2))})
ds = Dataset.from_dict({"a": [[[0, 0], [0, 0]]]}, features=features)
df = ds._data.to_pandas()
df.head()
#                a
# 0 [[0 0]
#  [0 0]]
df["a"].dtype
# <datasets.features.PandasArrayExtensionDtype at 0x17a2ed080>
df.to_csv("a.csv")
```
raises:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-7-df968361d1f9> in <module>
----> 1 df.to_csv("a.csv")

~/.virtualenvs/hf-datasets/lib/python3.7/site-packages/pandas/core/generic.py in to_csv(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, decimal)
   3202             decimal=decimal,
   3203         )
-> 3204         formatter.save()
   3205 
   3206         if path_or_buf is None:

~/.virtualenvs/hf-datasets/lib/python3.7/site-packages/pandas/io/formats/csvs.py in save(self)
    202             )
    203 
--> 204             self._save()
    205 
    206         finally:

~/.virtualenvs/hf-datasets/lib/python3.7/site-packages/pandas/io/formats/csvs.py in _save(self)
    321                 break
    322 
--> 323             self._save_chunk(start_i, end_i)
    324 
    325     def _save_chunk(self, start_i: int, end_i: int) -> None:

~/.virtualenvs/hf-datasets/lib/python3.7/site-packages/pandas/io/formats/csvs.py in _save_chunk(self, start_i, end_i)
    336                 decimal=self.decimal,
    337                 date_format=self.date_format,
--> 338                 quoting=self.quoting,
    339             )
    340 

~/.virtualenvs/hf-datasets/lib/python3.7/site-packages/pandas/core/internals/blocks.py in to_native_types(self, slicer, na_rep, quoting, **kwargs)
   1819 
   1820         values = np.asarray(values.astype(object))
-> 1821         values[mask] = na_rep
   1822 
   1823         # we are expected to return a 2-d ndarray

IndexError: arrays used as indices must be of integer (or boolean) type
```

It looks like the PandasArrayExtensionArray doesn't implement `isna` and so the mask can't be applied.

Reply:
I can open a PR to fix that

Reply:
Good to know, thanks!
I'll look into it

Reply:
I merged a fix for the custom features, exporting to csv for them now works as expected :)